napari/utils/_magicgui.py

```diff
@@ -0,0 +1,88 @@
+their custom classes or types should be converted into GUIs.  Then, when the
+end-user annotates one of their function arguments with a type hint using one
+of those custom classes, magicgui will know what to do with it.
+
+"""
+from typing import Tuple, Type, Any
+from ..layers import Layer
+
+try:
+    from magicgui import register_type
+
+    def get_layers(gui, layer_type: Type[Layer]) -> Tuple[Layer, ...]:
+        """Retrieve layers of type `layer_type`, from the Viewer the gui is in.
+
+        Parameters
+        ----------
+        gui : MagicGui or QWidget
+            The instantiated MagicGui widget.  May or may not be docked in a
+            dock widget.
+        layer_type : type
+            This is the exact type used in the type hint of the user's
+            function. It may be a subclass of napari.layers.Layer
+
+        Returns
+        -------
+        tuple
+            Tuple of layers of type ``layer_type``
+
+        Example
+        -------
+        This allows the user to do this, and get a dropdown box in their GUI
+        that shows the available image layers.
+
+        >>> @magicgui
+        ... def get_layer_mean(layer: napari.layers.Image) -> float:
+        ...     return layer.data.mean()
+
+        """
+        try:
+            # look for the parent Viewer based on where the magicgui is docked.
+            # if the magicgui widget does not have a parent, it is unattached
+            # to any viewers, and therefore we cannot return a list of layers
+            viewer = gui.parent().qt_viewer.viewer
+            return tuple(l for l in viewer.layers if isinstance(l, layer_type))
+        except AttributeError:
+            return ()
+
+    def show_result(gui, result: Any, return_type: Type[Layer]) -> None:
+        """Show a magicgui result in the viewer.
+
+        Parameters
+        ----------
+        gui : MagicGui or QWidget
+            The instantiated MagicGui widget.  May or may not be docked in a
+            dock widget.
+        result : Any
+            The result of the function call.
+        return_type : type
+            The return annotation that was used in the decorated function.
+        """
+        if result is None:
+            return
+
+        try:
+            viewer = gui.parent().qt_viewer.viewer
+        except AttributeError:
+            return
+
+        result_name = gui.func.__name__ + " result"
+        try:
+            viewer.layers[result_name].data = result
+        except KeyError:
+            adder = getattr(viewer, f'add_{return_type.__name__.lower()}')
+            adder(data=result, name=result_name)
+
+    # tell magicgui that whenever an argument is annotated as a
+    # napari.layers.Layer, then it should be rendered as a ComboBox,
+    # where the items are the current Viewer.layers of a certain type.
+    register_type(Layer, choices=get_layers, return_callback=show_result)
+
+except ImportError:
+    pass
```

Comment:
I generally dislike modules that affect global state on import. I also don't love that this whole module is wrapped in a try/except.

My suggestions would be:

- take the above function definitions out of the try/except
- use a try/except on the import, where `register_type` is a do-nothing function if the magicgui module isn't available.
- in the `napari/__init__.py`, after importing, run the function, rather than running it in this module.

I'm still a bit cautious about running the function automagically on importing napari, even, but at least we give people digging in a chance to quickly find out where this is happening.

Reply:
Thank @jni, these are great suggestions, and I incorporated them all.  I understand the concern about running the function on napari import... if it makes you feel better, all the `register_type` function is doing in `magicgui` is populating a `dict` in magicgui.  
for what it's worth, with magicgui installed in my environment adding this registration to `napari.__init__` did not add a detectable increase to napari import time using `time python -c "import napari"` (i.e. any increase was less than the standard deviation of import times)

Reply:
Thanks @tlambert03! My concern is not about import time, it's just that when I am debugging some behaviour, it never occurs to me that the bug might be coming from an *import* rather than a function running. So e.g. I typically will step right over the import lines and then be like "WTF HOW DID THIS DICTIONARY GET POPULATED", upend the table, etc. ;) Anyway, I think (before having run through the code again) that this will be in a satisfactory place at this point. =)

Reply:
hah, such a temper!  well, before we merge this, I'll just throw out three more possibilities:
1. no automagic behavior: we just create the `_magicgui.register_types_with_magicgui` function but don't call it, and anyone who wants to use `napari` with `magicgui` will just have to import the function from napari and call it themselves.
2. we put the `try: import napari` behavior in magicgui instead (I don't love that, since ultimately it'd be nice for other packages to support their own types, but it's definitely an option).
3. we stick with the current plan, but if anyone opens an issue because something unexpected happened and running this function was the culprit, we banish it!

Reply:
Yeah, option 3 is what I'd settled for, and happy to proceed with it. =)